File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    //sliding window combined with heights problem
    largestRectangleArea(heights) {
        let stack = [];//index and height
        let maxArea = 0;

        for(let i = 0; i< heights.length;i++){
            let start = i;
            while(stack.length > 0 && stack[stack.length -1][1] > heights[i]){
                const [index,height] = stack.pop();
                maxArea = Math.max(maxArea, height*(i-index));
                start = index;
            }
            stack.push([start,heights[i]]);
        }
        for(const [index,height] of stack){
            maxArea = Math.max(maxArea, height * (heights.length-index));
        }

        return maxArea;
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-0.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n + 1, vector<int>(2,0));
        dp[n-1][1] = dp[n-1][0]= nums[n-1];
        for(int i =n-2; i>=0;i--){
            dp[i][1]= max(nums[i],nums[i]+ dp[i+1][1]);
            dp[i][0]= max(dp[i+1][0], dp[i][1]);
        }

        return dp[0][0];
    }
};
